File: solutions/ocaml/dnd-character/1/dnd_character.ml
type character = {
  charisma : int;
  constitution : int;
  dexterity : int;
  hitpoints : int;
  intelligence : int;
  strength : int;
  wisdom : int;
}

let ability () =
  let values = List.init 4 Fun.id |> List.map (fun _ -> Random.int 6 + 1) in
  let sum = List.fold_left ( + ) 0 values in
  let min = List.fold_left Int.min Int.max_int values in
  sum - min

let modifier ~score =
  int_of_float @@ Float.floor @@ (float_of_int (score - 10) /. 2.0)

let generate_character () =
  let strength = ability () in
  let dexterity = ability () in
  let constitution = ability () in
  let intelligence = ability () in
  let wisdom = ability () in
  let charisma = ability () in
  let hitpoints = 10 + modifier ~score:constitution in
  {
    strength;
    dexterity;
    constitution;
    intelligence;
    wisdom;
    charisma;
    hitpoints;
  }

File: solutions/r/dnd-character/1/dnd-character.R
modifier <- function(score) {
  floor((score - 10) / 2)
}

ability <- function() {
  rolls <- sample(1:6, size = 4, replace = TRUE)
  # Sort descending and sum the top 3
  sum(sort(rolls, decreasing = TRUE)[1:3])
}

new_character <- function() {
  strength     <- ability()
  dexterity    <- ability()
  constitution <- ability()
  intelligence <- ability()
  wisdom       <- ability()
  charisma     <- ability()
  
  hitpoints    <- 10 + modifier(constitution)
  
  character <- list(
    strength     = strength,
    dexterity    = dexterity,
    constitution = constitution,
    intelligence = intelligence,
    wisdom       = wisdom,
    charisma     = charisma,
    hitpoints    = hitpoints
  )
  
  # Set class to both "character" and "list" to preserve S3 list behavior
  class(character) <- c("character", "list")
  return(character)
}